lib/ansible/modules/cloud/amazon/aws_kms.py

```diff
@@ -625,134 +607,173 @@ def compare_grants(existing_grants, desired_grants, purge_grants=False):
+    try:
+        new_policy = json.loads(policy)
+    except ValueError as e:
+        module.fail_json_aws(e, msg="Unable to parse new policy as JSON")
+
+    key_id = key['key_arn']
+    try:
+        keyret = connection.get_key_policy(KeyId=key_id, PolicyName='default')
         original_policy = json.loads(keyret['Policy'])
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError):
+        # If we can't fetch the current policy assume we're making a change
+        # Could occur if we have PutKeyPolicy without GetKeyPolicy
+        original_policy = {}
+
+    if not compare_policies(original_policy, new_policy):
+        return False
+
+    if not module.check_mode:
         try:
-            new_policy = json.loads(policy)
-        except ValueError as e:
-            module.fail_json_aws(e, msg="Unable to parse new policy as JSON")
-        if compare_policies(original_policy, new_policy):
-            changed = True
-            if not module.check_mode:
-                try:
-                    connection.put_key_policy(KeyId=key_id, PolicyName='default', Policy=policy)
-                except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-                    module.fail_json_aws(e, msg="Unable to update key policy")
+            connection.put_key_policy(KeyId=key_id, PolicyName='default', Policy=policy)
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Unable to update key policy")
+
+    return True
 
-    desired_grants = module.params.get('grants')
+
+def update_grants(connection, module, key, desired_grants, purge_grants):
     existing_grants = key['grants']
 
-    to_add, to_remove = compare_grants(existing_grants, desired_grants,
-                                       module.params.get('purge_grants'))
-    if to_remove:
+    to_add, to_remove = compare_grants(existing_grants, desired_grants, purge_grants)
+    if not (bool(to_add) or bool(to_remove)):
+        return False
+
+    key_id = key['key_arn']
+    if not module.check_mode:
         for grant in to_remove:
             try:
                 connection.retire_grant(KeyId=key_id, GrantId=grant['grant_id'])
-                changed = True
             except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
                 module.fail_json_aws(e, msg="Unable to retire grant")
-
-    if to_add:
         for grant in to_add:
             grant_params = convert_grant_params(grant, key)
             try:
                 connection.create_grant(**grant_params)
-                changed = True
             except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
                 module.fail_json_aws(e, msg="Unable to create grant")
 
+    return True
+
+
+def update_key(connection, module, key):
+    changed = False
+
+    changed |= cancel_key_deletion(connection, module, key)
```

Comment:
This function doesn't take check_mode into account

Reply:
ok, I'll get that fixed.